// Copyright 2023 Ant Group Co., Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package kuscia.proto.api.v1alpha1.kusciaapi;

import "kuscia/proto/api/v1alpha1/common.proto";

option go_package = "github.com/secretflow/kuscia/proto/api/v1alpha1/kusciaapi";
option java_package = "org.secretflow.v1alpha1.kusciaapi";

service DomainDataService {
  rpc CreateDomainData(CreateDomainDataRequest) returns (CreateDomainDataResponse);

  rpc UpdateDomainData(UpdateDomainDataRequest) returns (UpdateDomainDataResponse);

  rpc DeleteDomainData(DeleteDomainDataRequest) returns (DeleteDomainDataResponse);

  rpc QueryDomainData(QueryDomainDataRequest) returns (QueryDomainDataResponse);

  rpc BatchQueryDomainData(BatchQueryDomainDataRequest) returns (BatchQueryDomainDataResponse);

  rpc ListDomainData(ListDomainDataRequest) returns (ListDomainDataResponse);
}

message CreateDomainDataRequest {
    RequestHeader header = 1;
    // Optional, the domaindata_id would be generated by server if the domaindata_id is empty.
    // The unique identity of domaindata,it couldn't duplicate in the same domain.
    string domaindata_id = 2;
    // The human readable, it could duplicate in the domain.
    string name = 3;
    // Enum: table,model,rule,report,unknown
    string type = 4;
    // The relative_uri is relative to the datasource URI, The datasourceURI appends relative_uri is the domaindataURI.
    // e.g. the relative_uri is "train/table.csv"
    //      the URI of datasource is "/home/data"
    //      the URI of domaindata is "/home/data/train/table.csv"
    string relative_uri = 5;
    // Domain_id is the unique identity of the domain. the domaindata is belong to this domain.
    string domain_id = 6;
    // Optional, server would use default datasource if datasource_id is empty.
    // The datasource is where the domain is stored.
    string datasource_id = 7;
    // Optional, The attributes of the domaindata, this field use as a extra field, User could set
    // this field to any data what they need.
    map<string,string> attributes = 8;
    // Optional, Partition not support now
    Partition partition = 9;
    // This field must be set if the type is 'table',
    // the columns describe the table's schema information.
    repeated DataColumn columns = 10;
    // Optional , The vendor is the one who outputs the domain data, it may be the SecretFlow engine,
    // another vendor's engine, or manually registered. it's could be manual, secretflow or other vendor string.
    string vendor = 11;
    // file-format only takes effect when the data source type is  localfs or oss. default value is csv
    FileFormat file_format = 12;
}

message CreateDomainDataResponse {
  Status status = 1;
  CreateDomainDataResponseData data = 2;
}

message CreateDomainDataResponseData {
  // id of created datasource
  string domaindata_id = 1;
}

// UpdateDomainDataRequest : The field would not be updated if the field not set in the UpdateDomainDataRequest
// or the field is invalid.
message UpdateDomainDataRequest {
    RequestHeader header = 1;
    // Mandatory, The domaindata_id indicate which domaindata would be updated.
    string domaindata_id = 2;
    // The human-readable name
    string name = 3;
    // Enum: table,model,rule,report,unknown
    string type = 4;
    // The relative_uri is relative to the datasource URI, The datasourceURI appends relative_uri is the domaindataURI.
    // e.g. the relative_uri is "train/table.csv"
    //      the URI of datasource is "/home/data"
    //      the URI of domaindata is "/home/data/train/table.csv"
    string relative_uri = 5;
    // Mandatory, The domain_id indicate which domain's domaindata would be updated.
    string domain_id = 6;
    // The datasource is where the domain is stored.
    string datasource_id = 7;
    // The attributes of the domaindata, this field use as a extra field, User could set
    // this field to any data that they need.
    map<string,string> attributes = 8;
    // Partition not support now
    Partition partition = 9;
    // The columns describe the table's schema information.
    repeated DataColumn columns = 10;
    // Optional , The vendor is the one who outputs the domain data, it may be the SecretFlow engine,
    // another vendor's engine, or manually registered. it's could be manual, secretflow or other vendor string.
    string vendor = 11;
    // file-format only takes effect when the data source type is  localfs or oss. default value is csv
    FileFormat file_format = 12;
}

message UpdateDomainDataResponse {
  Status status = 1;
}

message DeleteDomainDataRequest {
  RequestHeader header = 1;
  // domaindata id
  string domain_id = 2;
  string domaindata_id = 3;
}

message DeleteDomainDataResponse {
  Status status = 1;
}

message QueryDomainDataRequest {
  RequestHeader header = 1;
  QueryDomainDataRequestData data = 2;
}

message QueryDomainDataResponse {
  Status   status = 1;
  DomainData data = 2;
}

message QueryDomainDataRequestData {
   string domain_id = 1;
   string domaindata_id = 2;
}

message BatchQueryDomainDataRequest {
  RequestHeader header = 1;
  repeated QueryDomainDataRequestData data = 2;
}

message BatchQueryDomainDataResponse {
  Status   status = 1;
  DomainDataList data = 2;
}

message ListDomainDataRequest {
  RequestHeader header = 1;
  ListDomainDataRequestData data = 2;
}

message ListDomainDataResponse {
  Status   status = 1;
  DomainDataList data = 2;
}

message ListDomainDataRequestData {
  // couldn't be empty
  string domain_id = 1;
  // Optional, The domaindata_type would be use as the filter condition to list the domaindata.
  string domaindata_type = 2;
  // Optional, the domaindata_vendor would be use as the filter condition to list the domaindata.
  string domaindata_vendor = 3;
}

message DomainDataList {
  repeated DomainData domaindata_list = 1;
}

message DomainData {
    // domaindata_id is the identification of domaindata, it couldn't duplicate in the same domain.
    string domaindata_id = 1;
    // The human readable, it could duplicate in the domain.
    string name = 2;
    // DomainData type , Enum: table,model,rule,report,unknown
    string type = 3;
    // The relative_uri is relative to the datasource URI, The datasourceURI appends relative_uri is the domaindataURI.
    // e.g. the relative_uri is "train/table.csv"
    //      the URI of datasource is "/home/data"
    //      the URI of domaindata is "/home/data/train/table.csv"
    string relative_uri = 4;
    // domain_id the unique identity of the domain. the domaindata is belong to this domain.
    string domain_id = 5;
    // datasource_id is the identity of the domaindatasource, the domaindatasource that storage the domaindata file.
    string datasource_id = 6;
    // The attributes of the domaindata, this field use as a extra field, User could set
    // this field to any data that they need.
    map<string,string> attributes = 7;
    // Partition not support now,
    Partition partition = 8;
    // This field must be set if the type is 'table',
    // The columns describe the table's schema information.
    repeated DataColumn columns = 9;
    // The vendor is the one who outputs the domain data, it may be the SecretFlow engine,
    // another vendor's engine, or manually registered. it's could be manual, secretflow or other vendor string.
    string vendor = 10;
    // The status of the domaindata , enum: Available,Unavailable
    string status = 11;
    string author = 12;
    // file-format only takes effect when the data source type is  localfs or oss. default value is csv
    FileFormat file_format = 13;
}